### ISSUE TITLE
Grad from `diffdist` differs from manual method

### ISSUE THREAD

**Issue by qizhuli** [ag14774/diffdist#1]
Hi thanks for the awesome repo! I found my way here from the [post](https://discuss.pytorch.org/t/how-to-preserve-backward-grad-fn-after-distributed-operations/49343/8) on pytorch forum. Curious how your implementation compares with manually assigning `tensor_i` to `tensor_list[i]` after `all_gather`, I extended the toy example in the forum post. My `manual.py` looks like this:
```python
import torch
import torch.distributed as dist

dist.init_process_group(backend="nccl", init_method="env://")
torch.cuda.set_device(dist.get_rank())
dist.barrier()
# in each process:
a = torch.tensor([1.0, 3.0], requires_grad=True, device='cuda')
b = a + 2 * dist.get_rank()
# gather
bs = [torch.empty_like(b) for i in range(dist.get_world_size())]
dist.all_gather(bs, b)
bs[dist.get_rank()] = b
# loss backward
loss = (torch.cat(bs) * torch.cat(bs)).mean()
loss.backward()
print(a.grad)
```
On the other hand, using your implementation, my `auto.py` contains the following code:
```python
import diffdist
import torch
import torch.distributed as dist

dist.init_process_group(backend="nccl", init_method="env://")
torch.cuda.set_device(dist.get_rank())
dist.barrier()
# in each process:
a = torch.tensor([1.0, 3.0], requires_grad=True, device='cuda')
b = a + 2 * dist.get_rank()
# gather
bs = [torch.empty_like(b) for i in range(dist.get_world_size())]
bs = diffdist.functional.all_gather(bs, b)
# loss backward
loss = (torch.cat(bs) * torch.cat(bs)).mean()
loss.backward()
print(a.grad)
```
Both scripts are identical except the method used for `all_gather`. So I expect the output to be identical too.
Now I launch both scripts using the following routine:
```sh
export CUDA_VISIBLE_DEVICES=0,1
python -m torch.distributed.launch --nproc_per_node=2 <script_name>
```
where `<script_name>` is either `manual.py` or `auto.py`. 

However, to my surprise, they produce different outputs. `manual.py` gives:
```py
tensor([0.5000, 1.5000], device='cuda:0')
tensor([1.5000, 2.5000], device='cuda:1')
```
whereas `auto.py` produces:
```py
tensor([1., 3.], device='cuda:0')
tensor([3., 5.], device='cuda:1')
```
It seems that the gradient from `diffdist` is double that of the manual method. Is this an intended effect? Also, can you shed any light on why this disagreement came about?

Many thanks again for sharing your repo!

(torch version 1.1.0)

**Comment by qizhuli** [comment]
Update: The scale factor between `torch.distributed` gradient and `diffdist` gradient seems to be exactly the number of processes spawned. 

I tried running the same script on three and four nodes. With three nodes, `manual.py` produces:
```
tensor([1., 3.], device='cuda:0')
tensor([3., 5.], device='cuda:1')
tensor([5., 7.], device='cuda:2')
```
and `auto.py` gives:
```
tensor([0.3333, 1.0000], device='cuda:0')
tensor([1.0000, 1.6667], device='cuda:1')
tensor([1.6667, 2.3333], device='cuda:2')
```
Clearly `diffdist` gradients are three times those of `torch.distributed`.

With four nodes, `auto.py` gives:
```
tensor([1., 3.], device='cuda:0')
tensor([3., 5.], device='cuda:1')
tensor([5., 7.], device='cuda:2')
tensor([7., 9.], device='cuda:3')
```
and `manual.py` gives:
```
tensor([0.2500, 0.7500], device='cuda:0')
tensor([0.7500, 1.2500], device='cuda:1')
tensor([1.2500, 1.7500], device='cuda:2')
tensor([1.7500, 2.2500], device='cuda:3')
```
Now `diffdist` gradients are four times those of `torch.distributed`.

**Comment by MetagenomicMIL** [comment]
Hello,
I've been busy and I've just noticed that there is an open issue. I'll look more into your example in more details soon.
Thanks!

**Comment by ag14774** [comment]
Hi,
I've only just noticed the open issue. I will have a more thorough look and if indeed there is a problem I will release a new version.
Thanks

**Comment by ag14774** [comment]
So your manual method doesn't take into account how tensor `a` affected the result in other processes. You gather all tensors in `bs`, then you manually set tensor `b` to the location indicated by the process rank. So only that particular element has a gradient function hooked into it

**Comment by ag14774** [comment]
```
a = torch.tensor([1.0, 3.0], requires_grad=True, device='cuda')
b = a + 2
# gather
bs = [torch.empty_like(b) for i in 4]  # Let's say we simulate 4 processes
bs[0] = b
bs[1] = torch.tensor([1.0, 3.0], device='cuda') + 2 * 1
bs[2] = torch.tensor([1.0, 3.0], device='cuda') + 2 * 2
bs[3] = torch.tensor([1.0, 3.0], device='cuda') + 2 * 3
# loss backward
loss = (torch.cat(bs) * torch.cat(bs)).mean()
loss.backward()
print(a.grad)
```

**Comment by ag14774** [comment]
My guess is that the code above will have the same result as your manual code. Just doing `dist.all_gather` does not connect the graph in process `i` with the graph of process `j`. So to the autograd engine of process `i`, it is like you are just manually setting all those positions to the correct value (except position `i`). So your example does not really simulate 4 processes.

Here is a link to a proper test for `all_gather`: https://github.com/ag14774/diffdist/blob/9a753a2e6e15ff662cecc9ae37273d39c3af737f/diffdist/testing.py#L20

**Comment by qizhuli** [comment]
Hi, thanks for the reply. I don't quite follow your logic though. How is `a` in process rank 0 supposed to affect what's going on in process rank 1, which is an independent process itself?

So I had done some math for the 2-node example. Using subscripts to indicate the tensor's device, we have:
```
a_0 = [1, 3], b_0 = [1, 3]
a_1 = [1, 3], b_1 = [3, 5]
```
Let's also use a second subscript to denote the index within the tensor. So `a_00 = 1` and `b_10 = 3`.
Now let's say we do an `all_gather` and produce variable `c` (i.e., the `bs` in the code snippet):
```
c = [[1, 3], [3, 5]]
```
Now we do what the code does to obtain the loss. Firstly, squaring `c` element-wise and putting the result into `d`, we have:
```
d = [[1, 9], [9, 25]]
```
Summing all elements in `d` we have:
```
s = 1 + 9 + 9 + 25 = 44
```
The final loss is the mean:
```
l = 1/4 * s = 11
```
Now we calculate loss. For now, let's focus on `∂l/∂a_10`. By chain rule, we have:
```
∂l/∂a_10 = ∂l/∂s * ∂s/∂d_10 * ∂d_10/∂c_10 * ∂c_10/∂b_10 * ∂b_10/∂a_10
```
These are the partial derivatives I have obtained:
```
∂l/∂s = 1/4
∂s/∂d_10 = 1
∂d_10/∂c_10 = 2 * c_10 = 6
∂c_10/∂b_10 = 1
∂b_10/∂a_10 = 1
```
Therefore, I got the final `∂l/∂a_10 = 1/4 * 6 = 1.5`, which agrees with the manual result. What's your view on this?

**Comment by ag14774** [comment]
But you don't have 1 loss. You have 2 losses. Same loss but you have it twice. If you change `a_0` it will have an effect on the loss in process 2 right? So when you backpropagate, you backpropagate twice. The gradient will accumulate

**Comment by ag14774** [comment]
So when you backpropagate, you backpropagate twice. The gradient will accumulate

**Comment by qizhuli** [comment]
You are right! In the `manual.py`, that is no gradient link between `c` and `b` residing on different devices/processes, whereas `auto.py` retains their gradient links. Thanks a lot for the explanation :)

**Comment by ag14774** [comment]
You are welcome!

If you want to have just one loss you actually need `gather` in order to gather everything in a single process. Unfortunately, NCCL does not currently support this so you have to use a different backend such as MPI.

I am closing the issue.
